### PR TITLE
Merging Terraform, App, Infrastrcture updates to Main

### PR DESCRIPTION
**Pull request by NarmathaBala**
This PR includes
1. Infrastructure updates using Terraform to create JMeter Distributed environment
2. Infrastructure updates to create VMSS, App Gateway and DSC configuration to set the VM in desired state
3. Integrate app creation and deployment code

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -22,293 +19,302 @@ parameters:
...
-            --logger trx
+          name: "Verify Azure install"
+          command: az -v
 
+      - azure-cli/login-with-service-principal:
+          azure-sp: ARM_CLIENT_ID
+          azure-sp-password: ARM_CLIENT_SECRET
+          azure-sp-tenant: ARM_TENANT_ID
+    
+
```
Comment (phongcao) [inline .circleci/config.yml:67]: nit: redundant line

File: .circleci/config.yml [hunk 2]
```diff
@@ -22,293 +19,302 @@ parameters:
...
+      - azure-cli/login-with-service-principal:
+          azure-sp: ARM_CLIENT_ID
+          azure-sp-password: ARM_CLIENT_SECRET
+          azure-sp-tenant: ARM_TENANT_ID
+  
+  login-to-azure-down:
     steps:
-      - checkout
+      - azure-cli/install
+      - run: sudo apt-cache policy azure-cli | sudo apt-get install --allow-downgrades azure-cli=2.18.0-1~stretch
```
Comment (phongcao) [inline .circleci/config.yml:57]: Why do we need to downgrade az-cli?

File: .circleci/config.yml [hunk 3]
```diff
@@ -22,293 +19,302 @@ parameters:
...
-            trx2junit << pipeline.parameters.test_dir >>/TestResults/*.trx
+            sudo apt-get update && sudo apt-get install -y gnupg software-properties-common curl
+            curl -fsSL https://apt.releases.hashicorp.com/gpg | sudo apt-key add -
+            sudo apt-add-repository "deb [arch=amd64] https://apt.releases.hashicorp.com $(lsb_release -cs) main"
+            sudo apt-get update && sudo apt-get install terraform
 
-      - store_test_results:
-          name: "Store test results"
-          path: << pipeline.parameters.test_dir >>/TestResults
+  install-azcopy-linux:
```
Comment (phongcao) [inline .circleci/config.yml]: Should we remove this command since there is no use for it anymore?

File: .circleci/config.yml [hunk 4]
```diff
@@ -22,293 +19,302 @@ parameters:
...
-            az login `
-            --service-principal `
-            --tenant "$Env:TENANT_ID" `
-            --username "$Env:CLIENT_ID" `
-            --password "$Env:CLIENT_SECRET"
-
-            # Set default subscription
-            az account set `
-            --subscription "$Env:SUBSCRIPTION_ID"
+            cd scripts
```
Comment (phongcao) [inline .circleci/config.yml]: Better to use `working_directory: scripts` with the run step.

File: .circleci/config.yml [hunk 5]
```diff
@@ -22,293 +19,302 @@ parameters:
...
-            --name "$func" `
-            --src << pipeline.parameters.function_zip_file >>
-
             # Output variables to file
-            echo "`$buildNum=`"$Env:CIRCLE_BUILD_NUM`"" >> output.ps1
-            echo "`$targetHost=`"$func.azurewebsites.net`"" >> output.ps1
+            # echo "TF_VAR_TARGET_HOSTNAME=$func.azurewebsites.net" >> .env
+            # echo "target_hostname=$func.azurewebsites.net" >> << pipeline.parameters.properties_file_name >>
+            echo "TF_VAR_TARGET_HOSTNAME=microsoft.com" >> .env
+            echo "target_hostname=microsoft.com" >> << pipeline.parameters.properties_file_name >>
```
Comment (phongcao) [inline .circleci/config.yml:120]: Are we going to update the hostname in the next integration?

File: .circleci/config.yml [hunk 6]
```diff
@@ -22,293 +19,302 @@ parameters:
...
-            # Deploy zip file
-            az functionapp deployment source config-zip `
-            --resource-group "$rg" `
-            --name "$func" `
-            --src << pipeline.parameters.function_zip_file >>
-
             # Output variables to file
-            echo "`$buildNum=`"$Env:CIRCLE_BUILD_NUM`"" >> output.ps1
-            echo "`$targetHost=`"$func.azurewebsites.net`"" >> output.ps1
+            # echo "TF_VAR_TARGET_HOSTNAME=$func.azurewebsites.net" >> .env
```
Comment (phongcao) [inline .circleci/config.yml]: Can you remove those lines since we don't use them anymore?

File: .circleci/config.yml [hunk 7]
```diff
@@ -22,293 +19,302 @@ parameters:
...
-            --name "rg-loadtest$buildNum" `
-            --yes
+      - store_test_results:
+          name: "Store test results"
+          path: terraform
+
+      - store_artifacts:
+          path: terraform/<< pipeline.parameters.junit_file_name >>
+
+      # - terraform/destroy:
```
Comment (phongcao) [inline .circleci/config.yml]: Should we uncomment this step or just remove it?

File: .circleci/config.yml [hunk 8]
```diff
@@ -22,293 +19,302 @@ parameters:
...
+      - login-to-azure-down
+
+      - run:
+          name: "Clean up environment"
+          command: |
+            cd scripts
+            chmod a+x ./cleanup.sh
+            ./cleanup.sh << pipeline.parameters.environment_name >> jmeter
+
+
```
Comment (phongcao) [inline .circleci/config.yml]: nit: redundant line

File: .circleci/config.yml [hunk 9]
```diff
@@ -22,293 +19,302 @@ parameters:
...
+      - attach_workspace:
+          at: .
+      
+      - login-to-azure-down
+
+      - run:
+          name: "Clean up environment"
+          command: |
+            cd scripts
```
Comment (phongcao) [inline .circleci/config.yml]: Better to use `working_directory: scripts` with the run step.

File: .circleci/config.yml [hunk 10]
```diff
@@ -22,293 +19,302 @@ parameters:
...
           type: approval
+      - build_image:
           requires:
-            - package
-      - deploy:
-          requires:
-            - hold
-      - loadtest:
+            - build_image_approval
+        # filters:
```
Comment (phongcao) [inline .circleci/config.yml]: These should be uncommented when we're about to merge to main.

File: .circleci/config.yml [hunk 11]
```diff
@@ -22,293 +19,302 @@ parameters:
...
+        #   branches:
+        #     only:
+        #       - main
+
+  build_deploy_loadtest:
+    jobs:
+      # - deploy_approval:
```
Comment (phongcao) [inline .circleci/config.yml]: These should be uncommented when we're about to merge to main.

File: .circleci/config.yml [hunk 12]
```diff
@@ -22,293 +19,302 @@ parameters:
...
+            - deploy_script
+      - cleanup_approval:
+          type: approval
           requires:
-            - deploy
+              - load_test
+      - cleanup:
+          requires: 
+            - cleanup_approval
+
```
Comment (phongcao) [inline .circleci/config.yml]: nit: so many redundant lines below

File: scripts/cleanup.sh [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+    echo "#### Deleting resource group: ${resource_group} ####"
+    az group delete \
+        --name "$resource_group" \
+        --yes -y
+    
+    echo "#### Deleting resource group: ${jmeter_resource_group} ####"
+    az group delete \
+        --name "$jmeter_resource_group" \
+        --yes -y
+
```
Comment (phongcao) [inline scripts/cleanup.sh]: nit: redundant line

**Review by phongcao — CHANGES_REQUESTED** [review]

File: .circleci/config.yml (hunk 5 above)
Comment (nexilus18) [reply .circleci/config.yml:120]: Yes

File: .circleci/config.yml (hunk 3 above)
Comment (nexilus18) [reply .circleci/config.yml]: I removed it.

File: .circleci/config.yml (hunk 2 above)
Comment (nexilus18) [reply .circleci/config.yml:57]: There's a bug in the latest version of az-cli that is causing an error when creating the Bastion host.

Comment (NarmathaBala) [reply .circleci/config.yml:57]: https://github.com/Azure/azure-cli/pull/20305

**Review by phongcao — APPROVED** [review]

**Pull request merged**